pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -16,41 +16,187 @@ package packing
 
 import (
 	"context"
+	"fmt"
+	"sort"
 
 	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
 	"go.uber.org/zap"
 	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
 )
 
-type PodPacker struct {
+type podPacker struct {
+	// TODO use this ec2 API to get the instance types
 	ec2 ec2iface.EC2API
 }
 
-// PodPacker helps pack the pods and calculates efficient placement on the instances.
+// Packer helps pack the pods and calculates efficient placement on the instances.
 type Packer interface {
-	Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error)
+	// TODO use ctx when calling ec2 API
+	Pack(ctx context.Context, pods []*v1.Pod) ([]*Packing, error)
 }
 
-// Packings contains a list of pods that can be placed on any of Instance type
-// in the InstanceTypeOptions
-type Packings struct {
-	Pods                []*v1.Pod
-	InstanceTypeOptions []string
+// Packing contains a list of pods that can be placed on any of Instance type
+// in the InstanceTypes
+type Packing struct {
+	Pods          []*v1.Pod
+	InstanceTypes []string
 }
 
-func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
-	return &PodPacker{ec2: ec2}
+// NewPacker returns a Packer implementation
+func NewPacker(ec2 ec2iface.EC2API) Packer {
+	return &podPacker{ec2: ec2}
 }
 
 // Pack returns the packings for the provided pods. Computes a set of viable
 // instance types for each packing of pods. Instance variety enables EC2 to make
-// better cost and availability decisions.
-func (p *PodPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
-	zap.S().Debugf("Successfully packed %d pod(s) onto %d node(s)", len(pods), 1)
-	return []*Packings{
-		{
-			InstanceTypeOptions: []string{"m5.large"}, // TODO, prioritize possible instance types
-			Pods:                pods,
-		},
-	}, nil
+// better cost and availability decisions. Pods provided are all schedulable in
+// the same zone as tightly as possible. It follows the First Fit Decreasing bin
+// packing technique, reference-
+// https://en.wikipedia.org/wiki/Bin_packing_problem#First_Fit_Decreasing_(FFD)
+func (p *podPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packing, error) {
+	// TODO use ctx when calling ec2 API
+	return p.packPods(pods)
```

Comment:
why the indirection? If no other logic, I'd just collapse to one function.

Reply:
I actually had sorting logic before but removed it and I left it like this to unit test just the `packPods`

Reply:
For the unit testing could you change to test `Pack` instead?